realtek: Use constants from <linux/mii.h> and <linux/mdio.h>

hardfalcon opened the pull request:
Replace various numerical values, `BIT(x)` and `(1 << x)` constructs in `drivers/net/phy/rtl83xx-phy.c` with constants from `<linux/mii.h>` and `<linux/mdio.h>`.

Suggested-by: DENG Qingfang <dqfext@gmail.com>
Signed-off-by: Pascal Ernster <git@hardfalcon.net>

hardfalcon commented:
While writing this, I've wondered about the usage of `&= ~` in lines [2755](https://github.com/openwrt/openwrt/pull/12753/files#diff-1fa3d5314f13683d8f3a5cac5a841e2818865b0cbe1acbb164e31527e8406bbdL2755) and [2762](https://github.com/openwrt/openwrt/pull/12753/files#diff-1fa3d5314f13683d8f3a5cac5a841e2818865b0cbe1acbb164e31527e8406bbdL2762).

Shouldn't those lines use `|=` instead of `&= ~` to enable the phy, @svanheule?

LGA1150 commented:
I suggest use the pre-defined macros in `<linux/mii.h>`.

`PHY_CTRL_REG` -> `MII_BMCR`
`PHY_POWER_BIT` -> `BMCR_PDOWN`

svanheule commented:
> While writing this, I've wondered about the usage of `&= ~` in lines [2755](https://github.com/openwrt/openwrt/pull/12753/files#diff-1fa3d5314f13683d8f3a5cac5a841e2818865b0cbe1acbb164e31527e8406bbdL2755) and [2762](https://github.com/openwrt/openwrt/pull/12753/files#diff-1fa3d5314f13683d8f3a5cac5a841e2818865b0cbe1acbb164e31527e8406bbdL2762).
> 
> Shouldn't those lines use `|=` instead of `&= ~` to enable the phy, @svanheule?

The original name hints at a "power enable flag", but if it's actually a "power disable flag", then clearing it would be the right thing to do. Taking into account @LGA1150's remark, clearing should be correct. What certainly wouldn't make sense, is if the bit set/clear operations are not consistent with power-off/power-on throughout the code.

svanheule reviewed (commented):
Thanks for the clean-up.

As per the other suggestion, naming of the define-s could also be improved to prevent future confusion.

hardfalcon commented:
Thanks for the feedback and the suggestions, I've updated the PR accordingly.

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -749,7 +747,7 @@ static int rtl8380_configure_int_rtl8218b(struct phy_device *phydev)
 	// }
 
 	val = phy_read(phydev, 0);
```
This can be also replaced with `MII_BMCR`.

I actually took a good look at this file, and found that there are too many magic numbers which can be replaced with standard macros. May use a separate commit.

hardfalcon replied:
Thanks a lot for the hint, I've now also replaced the `0` with `MII_BMCR` in all other occurences of `phy_read(phydev, 0)` and `phy_write(phydev, 0, val)`.

svanheule commented:
Given that you haven't added a SoB to the last patch, I assume you still plan to clean up these changes?

hardfalcon commented:
@svanheule: I had planned to try to find a few more magical values to replace over the weekend, but didn't get around to do so.

Btw, should I keep these as 3 separate commits, or should I squash them into a single commit? Personally, I'd be inclined to squash them because they're all essentially doing the same type of thing within the same 1-2 files, but I'd be interested in the opinion of you folks on this.

svanheule commented:
No worries on the delay.

If these are all in one commit, it becomes a bit harder to spot mistakes. If you can find a logical split (e.g. one for the BMCR, one for the MMD registers) that helps maintaining a mental overview of what's supposed to happen, that would be preferable. But I don't think it's necessary to do one patch per replacement.

hardfalcon committed:
realtek: Use MII_BMCR and BMCR_* constants from <linux/mii.h>

Replace numerical values, BIT(x) and (1 << x) in
drivers/net/phy/rtl83xx-phy.c with constants from <linux/mii.h> to
improve code readability.

To make reviewing easier, this commit only addresses MII_BMCR and BMCR_*
constants.

Suggested-by: DENG Qingfang <dqfext@gmail.com>
Signed-off-by: Pascal Ernster <git@hardfalcon.net>

hardfalcon committed:
realtek: Use ADVERTISE_* and MII_PHYSID* from <linux/mii.h>

Replace BIT(x) and numerical values in drivers/net/phy/rtl83xx-phy.c
with constants from <linux/mii.h> to improve code readability.

To make reviewing easier, this commit only addresses ADVERTISE_* and
MII_PHYSID* constants.

Signed-off-by: Pascal Ernster <git@hardfalcon.net>

hardfalcon committed:
realtek: Use MDIO_* constants from <linux/mdio.h>

To improve code readability in drivers/net/phy/rtl83xx-phy.c, replace
constants MMD_AN and MMD_VEND2 from drivers/net/phy/rtl83xx-phy.h with
MDIO_MMD_AN and MDIO_MMD_VEND2 from <linux/mdio.h>.

Also, replace
BIT(0) with MDIO_EEE_2_5GT,
BIT(1) with MDIO_EEE_100TX,
BIT(2) with MDIO_EEE_1000T,
BIT(9) with MDIO_AN_CTRL1_RESTART,
BIT(12) with MDIO_AN_CTRL1_ENABLE,
32 with MDIO_AN_10GBT_CTRL,
60 with MDIO_AN_EEE_ADV, and
62 with MDIO_AN_EEE_ADV2
from <linux/mdio.h>.

Suggested-by: DENG Qingfang <dqfext@gmail.com>
Signed-off-by: Pascal Ernster <git@hardfalcon.net>

hardfalcon commented:
Sorry for the endless updates, but I think I'm finished at last. :)

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -493,34 +492,34 @@ static int rtl8226_advertise_aneg(struct phy_device *phydev)
 
 	/* Allow 2.5G */
-	v = phy_read_mmd(phydev, MMD_AN, 32);
+	v = phy_read_mmd(phydev, MDIO_MMD_AN, 32);
```
32 -> `MDIO_AN_10GBT_CTRL`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -493,34 +492,34 @@ static int rtl8226_advertise_aneg(struct phy_device *phydev)
 	if (v < 0)
 		goto out;
 
 	v |= BIT(7);
```
`BIT(7)` -> `MDIO_AN_10GBT_CTRL_ADV2_5G`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -537,22 +536,22 @@ static int rtl8226_config_aneg(struct phy_device *phydev)
 			goto out;
 		/* AutoNegotiationEnable */
-		v = phy_read_mmd(phydev, MMD_AN, 0);
+		v = phy_read_mmd(phydev, MDIO_MMD_AN, 0);
```
0 -> `MDIO_CTRL1`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -537,22 +536,22 @@ static int rtl8226_config_aneg(struct phy_device *phydev)
 
-		v |= BIT(12); /* Enable AN */
-		ret = phy_write_mmd(phydev, MMD_AN, 0, v);
+		v |= BMCR_ANENABLE;
```
Although it has the same value, it is preferred to use `MDIO_AN_CTRL1_ENABLE` for MMDs.

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -569,11 +568,11 @@ static int rtl8226_get_eee(struct phy_device *phydev,
 	pr_debug("In %s, port %d, was enabled: %d\n", __func__, addr, e->eee_enabled);
 
-	val = phy_read_mmd(phydev, MMD_AN, 60);
+	val = phy_read_mmd(phydev, MDIO_MMD_AN, 60);
```
60 -> `MDIO_AN_EEE_ADV`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -569,11 +568,11 @@ static int rtl8226_get_eee(struct phy_device *phydev,
 		e->eee_enabled = !!(val & BIT(1));
 		if (!e->eee_enabled) {
-			val = phy_read_mmd(phydev, MMD_AN, 62);
+			val = phy_read_mmd(phydev, MDIO_MMD_AN, 62);
```
62 -> `MDIO_AN_EEE_ADV2`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -594,29 +593,29 @@ static int rtl8226_set_eee(struct phy_device *phydev, struct ethtool_eee *e)
-	val = phy_read_mmd(phydev, MMD_AN, 60);
+	val = phy_read_mmd(phydev, MDIO_MMD_AN, 60);
 	if (e->eee_enabled)
 		val |= 0x6;
```
0x6 -> `MDIO_EEE_100TX | MDIO_EEE_1000T`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -569,11 +568,11 @@ static int rtl8226_get_eee(struct phy_device *phydev,
-	val = phy_read_mmd(phydev, MMD_AN, 60);
+	val = phy_read_mmd(phydev, MDIO_MMD_AN, 60);
 	if (e->eee_enabled) {
 		e->eee_enabled = !!(val & BIT(1));
```
`BIT(1)` -> `MDIO_EEE_100TX`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -569,11 +568,11 @@ static int rtl8226_get_eee(struct phy_device *phydev,
 		if (!e->eee_enabled) {
-			val = phy_read_mmd(phydev, MMD_AN, 62);
+			val = phy_read_mmd(phydev, MDIO_MMD_AN, 62);
 			e->eee_enabled = !!(val & BIT(0));
```
`BIT(0)` -> `MDIO_EEE_2_5GT`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -594,29 +593,29 @@ static int rtl8226_set_eee(struct phy_device *phydev, struct ethtool_eee *e)
-	val = phy_read_mmd(phydev, MMD_AN, 62);
+	val = phy_read_mmd(phydev, MDIO_MMD_AN, 62);
 	if (e->eee_enabled)
 		val |= 0x1;
```
`MDIO_EEE_2_5GT`

LGA1150 commented on target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -1057,12 +1056,11 @@ void rtl8218d_eee_set(struct phy_device *phydev, bool enable)
+	an_enabled = val & BMCR_ANENABLE;
 
-	/* Enable 100M (bit 1) / 1000M (bit 2) EEE */
 	val = phy_read_mmd(phydev, 7, 60);
```
`MDIO_MMD_AN, MDIO_AN_EEE_ADV`

hardfalcon commented:
Thanks a lot for the feedback and the suggestions, @LGA1150! :)

svanheule commented:
@LGA1150, feel free to leave a Reviewed-by: once you are happy with the changes.

svanheule commented:
Sorry for the delay. Merged to main!

87a2f03b8658 realtek: Use MDIO_* constants from <linux/mdio.h>
171e67e2f792 realtek: Use ADVERTISE_* and MII_PHYSID* from <linux/mii.h>
8b2f654d4ca4 realtek: Use MII_BMCR and BMCR_* constants from <linux/mii.h>
